File: services/app/dockers/js/checker.js
const readline = require("readline");
const chai = require("chai");

const assert = chai.assert;

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

const checks = [];

rl.on("line", function(line) {
  // console.log(line);
  checks.push(JSON.parse(line));
});

rl.on("close", function() {
  try {
    const solution = require("./check/solution");

    checks.forEach(function(check) {
      if (check.check) {
        process.stdout.write(JSON.stringify({
          status: "ok",
          result: check.check
        }));
      } else {
        const result = solution.apply(null, check.arguments);

        try {
          assert.deepEqual(result, check.expected);
        } catch (e) {
          process.stdout.write(JSON.stringify({
            status: "failure",
            result: check.arguments
          }));
          process.exit(1);
        }
      }
    });
  } catch (e) {
    process.stdout.write(JSON.stringify({
      status: "error",
      result: e.message
    }));
    process.exit(1);
  }
});

File: services/app/dockers/clojure/checker.clj
(ns checker
  (:require
    [clojure.test :refer :all]
    [clojure.data.json :as json]))

(try
  (load-file "check/solution.clj")
  (catch Exception ex
    (do
     (println (json/write-str {:status "error" :result (.getMessage ex)}))
     (System/exit 0)
     )))

(def data (line-seq (java.io.BufferedReader. *in*)))

(def prepared-data
  (json/read-str (str "[" (apply str data) "]") :key-fn keyword))

(defn generate-tests
  [data solution]
  (doseq [x data]
    (if (:check x)
      (println (json/write-str {:status "ok" :result (:check x)}))
      (try (assert (= (:expected x) (apply solution (:arguments x))))
           (catch java.lang.AssertionError e
             (do
              (println (json/write-str {:status "failure" :result (:arguments x)}))
              (System/exit 0)
              ))))))
(defn -main []
  (generate-tests prepared-data solution))
